### ISSUE TITLE
Out-String should not append a trailing newline

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#14444]
## Summary of the new feature/enhancement

It should be possible to use `Out-String` in a manner that doesn't _append a trailing newline_, which is what it currently does and has always done.

Note that `-NoNewLine` is _not_ a solution, because it suppresses _all_ newlines, so that _two or more_ output lines are then simply directly concatenated (like `-join ''`).

Here's a demonstration of the problem:

```powershell
# CURRENT behavior:
PS> ('foo' | Out-String) -replace '\r?\n', '<newline>'
foo<newline>   #  A newline was appended.
```

I see two options:

* (a) Consider a breaking change - assuming it is justified to consider it a [Bucket 3: Unlikely Grey Area](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-3-unlikely-grey-area) change - and make `Out-String` _never_ append a trailing newline.

* (b) Introduce a new switch, `-NoTrailingNewLine`

**Comment by bpayette** [comment]
`Out-String` formats an object for _presentation_ as a string. It *should* add a new line because that's how formatting works. Given that, I don't understand what the problem is - can you clarify? Thanks.

**Comment by mklement0** [comment]
>  It _should_ add a new line

I don't think it should. Why do you think so?

I would expect:

```powershell
Get-Date
```

and

```powershell
Get-Date | Out-String
``` 

to produce the same output, yet they don't (the latter has an extra newline).

Given that piping to `Out-String` is a convenient way to capture stdout from an _external program_ in a _single string_ - and remember that _strings_ are out-of-band objects with respect to formatting and are _passed through_ - it is inconvenient to have a trailing newline in the resulting string with respect to _further programmatic processing_.

In other words:

```powershell
PS> $out = cmd /c 'echo a&echo b' | Out-String
```

should be equivalent to:

```powershell
PS>  $out = (cmd /c 'echo a&echo b') -join [Environment]::NewLine
```

yet it isn't, due to the trailing newline.

**Comment by bpayette** [comment]
> there's no good reason to append a newline to the (formatted representation) of the last input object.

The purpose of `Out-String` is to render objects as presentable text which includes trailing newlines. If what you actually want to do is join a bunch of strings then use `-join` because that's it for.

**Comment by mklement0** [comment]
Yes, `-join` is a _workaround_, but that still doesn't justify the _trailing_ newline:

* If you _output_ the string resulting from `Out-String`, it is the standard output formatting that will add the trailing newline for you - just like sending the output to a _file_ will do that for you - just like interactively outputting `'foo'` prints a newline before the next prompt is displayed.

* If you _programmatically process_ `Out-String` output, the trailing newline is an unnecessary nuisance.

Each formatted output _individually_ may or may not have trailing newlines (e.g., `Get-Date` vs. `'foo'`).

Newlines are only needed as _separators between_ multiple formatted representations, not as a _terminators_.

**Comment by ExE-Boss** [comment]
As I said in <https://github.com/PowerShell/PowerShell/issues/13071#issuecomment-760563735>:
> This is the expected behaviour, because both <code>[Out‑String][out-string]</code> and <code>[Set‑Content][set-content]</code> add a trailing newline by default.
> 
> [out-string]: https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/out-string
> [set-content]: https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.management/set-content
> 
> ---
> 
> You can use the `‑NoNewline` switch with <code>[Set‑Content][set-content]</code> to disable adding the extra trailing newline:
> ```pwsh
> @(
> 	\"$($PSStyle.Foreground.Blue)Hello$($PSStyle.Reset)\",
> 	'string2'
> ) `
> 	| Out-String `
> 	| Set-Content -Path test2.txt -NoNewline
> ```

**Comment by mklement0** [comment]
@ExE-Boss: `Out-String` is an _in-memory_ operation, whereas `Set-Content` is a _file_ operation.

When you use `Out-String`, you typically want to process the result _in memory_ - otherwise, you'd just use `>` / `Out-File` directly.

For _in-memory_ processing, it is reasonable for multiple formatted representations to be joined with a _separator_ (a newline), and  not for each representation to have a _terminator_, which is a concept that applies to  (line-oriented text) _files_, especially on Unix.

If `Out-String` behaved that way, you could later use it with `Set-Content` _as-is_, without having to resort to the `-NoNewline` _workaround_.

**Comment by kilasuit** [comment]
Isn't the crux of this that in interactive output in the terminal shows a trailing new line and when run in a script (where there often isn't an interactive session) it shouldn't?

If so then that can be checked within the cmdlet whether or not the session is in non-interactive mode or not & I don't believe would be a breaking change

**Comment by mklement0** [comment]
@kilasuit, this is primarily about the trailing newline being unhelpful in _programmatic processing_ of `Out-String` output.

As for the display side of things:

Both default host output and file-creation cmdlets (`Out-File` / `>`, `Set-Content`) are clearly designed _not_ to expect the data to already have a trailing newline, and therefore _blindly append one_ (by default).
(And I don't see how interactive vs. script processing figures into this, but perhaps you can elaborate.)

Given that `Out-String` currently _itself_ appends a trailing newline to the resulting in-memory string, you end up having to fight that design:

* If you later split the string into lines, you'll end up with an extra empty array element at the end (``("foo`nbar`n" -split '\n').Count`` is 3).

* On printing to the host / saving to a file, you must then suppress the extra trailing newline:
  * With `-NoNewLine`, if you use `Out-File` / `Set-Content` (`>` is then  not an option)
  * For `Host-Output` you're stuck with something awkward like `Write-Host -NoNewline (Get-Date | Out-String)`

Of course, you can avoid the empty array element / the need for these workarounds by stripping the trailing newline manually first - (`(Get-Date | Out-String) -replace '\n\z'`) - but the whole point of this issue is that the trailing newline shouldn't be added to the in-memory data (string) to begin with: there is no upside to doing so, only downsides.

**Comment by ExE-Boss** [comment]
`Out‑String` also supports `‑NoNewline`.

**Comment by mklement0** [comment]
@ExE-Boss, to quote from the OP:

> Note that `-NoNewLine` is _not_ a solution, because it suppresses _all_ newlines, so that two or more output lines are then simply directly concatenated (like `-join ''`).

For instance:

```powershell
PS> 1, 2 | Out-String -NoNewline
12
```

**Comment by jazzdelightsme** [comment]
I think maybe mklement0 chose `Get-Date` as an example because it produces a simple object that *isn't* a string (a `DateTime`), so that `Out-String` would be doing "some work" to convert something to a string.

But consider even just a plain string!

```powershell
'asdf'
```

versus

```powershell
'asdf' | Out-String
```

How would one could consider that `'asdf'` is *not* "rendered as presentable text"? In other words, who would consider the output of the latter to be "presentable text", but *not* the former?

Adding newlines in is ONLY interesting where there are multiple input objects: newlines in that case are used to separate them (unless belayed by `-NoNewline`). The current last (trailing) newline is separating the last object from ... what? It serves no purpose; it is not separating it from anything. We should get rid of it.

Consider also this pathological scenario:

```powershell
'asdf' | Out-String | Out-String | Out-String | Out-String | Out-String 
```

Let's stop the madness and pull this burr out from under the saddle.

**Comment by SteveL-MSFT** [comment]
`Out-String` already has a `-NoNewline` switch I believe for this specific reason

**Comment by BobbyTables10** [comment]
Any updates on this? I'm using Out-String to generate an .xml (.nfo) file that will be picked up by Kodi to show movie details that I've scraped from an online video archive. Building up the .xml I use:

`   "    <title>" | Out-File -LiteralPath $FileName  -Encoding "UTF8" -Append `

.. then insert the scraped title. What I want to end up with in the .xml file is:

```
<movie>
<title>Some Movie Title</title>
<plot>The story begins in 1939.
A soldier ... 
</plot>
</movie>
```

but what I actually end up with is something like this, which embeds newlines into the title/plot text and looks scruffy when presented in the media centre.

<movie>
<title>
Some Movie Title

</title>
<plot>
The story begins in 1939.
A soldier ... 
</plot>
</movie>
```

As others have said, a `-NoTrailingNewLine ` switch would alleviate this without having to do loads of string processing?

**Comment by mklement0** [comment]
@BobbyTables10, the `-NoTrailingNewLine` / `-NoLastNewLine` proposal that also covers `Out-String`  is in #5108 - but, as argued above, the much preferable option is to simply omit a trailing newline from `Out-String`'s output to begin with.

What you're showing is an `Out-File` call; however, as with `Out-String`, if there's only _one_ input object (string), `-NoNewLine` does work as expected to suppress a trailing newline.
.

**Comment by jaydeepmehta15** [comment]
Upvote on -NoTrailingNewLine / -NoLastNewLine

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
Is "[System.Console]::Write" and optional manual addition of "[System.Environment]::Newline" inadequate?

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
http://stackoverflow.com/a/21564398/9731176

**Comment by ImportTaste** [comment]
> Is `[System.Console]::Write()` and optional manual addition of `[System.Environment]::Newline` inadequate? I may have misunderstood why Out-String exists, although it appears to me to be useless.
> 
> As demonstration, `[System.Console]::Write($(get-date))` or `[System.Console]::Write($(get-date),[System.Environment]::Newline)` appear perfect.

This may be incorrect, but I believe that Out-String does special handling of some PSObjects to ensure the output looks more-or-less alike to what it would've been without Out-String. Having a parameter like `-NoTrailingNewline` (perhaps with a `-ntn` alias) would be ideal.

**Comment by ImportTaste** [comment]
> Is `[System.Console]::Write()` and optional manual addition of `[System.Environment]::Newline` inadequate? I may have misunderstood why Out-String exists, although it appears to me to be useless.
> 
> As demonstration, `[System.Console]::Write($(get-date))` or `[System.Console]::Write($(get-date),[System.Environment]::Newline)` appear perfect.

Also, I just noticed today, these methods write to host rather than stdout.

**Comment by mklement0** [comment]
Nice, @Succubyss. Quibble: It's awkward to have a `[switch]` parameter that de facto defaults to `$true` - perhaps use `-AddTrailingNewline`?
As an aside that may be of interest: the proxy-function generation code is currently suboptimal - see https://github.com/PowerShell/PowerShell/issues/10863

**Comment by SteveL-MSFT** [comment]
Let me bring this up to the cmdlets WG.  I think we can use `EndProcessing` to determine the last item in the pipe and not append the final trailing newline.  Just need to discuss and think about unintended side effects.

**Comment by doctordns** [comment]
The Cmdlewt working group has reviewed this issue. We have decided changing the default behavior would break way too many things. Work rounds for things that don't like trailing line breaks have be around for a long time.  Resolved as  by design.

**Comment by jhoneill** [comment]
> > The Cmdlet working group has reviewed this issue. We have decided changing the default behavior would break way too many things. Work rounds for things that don't like trailing line breaks have be around for a long time. Resolved as by design.
> 
> A new switch to strip the trailing new line wouldn't break anything, and is a suggested resolution in the OP.

The issue is more complicated than posed in the OP.  (1) Formatting has never been consistent. A date gets leading and trailing line breaks (because it has formatting rules) but numbers and strings (with no formatting rules don't)
```
#613 [——€] PS7 ~\Documents\PowerShell> 1,2
1
2
#614 [——€] PS7 ~\Documents\PowerShell> $d=get-date; $d, $d.AddDays(1)

17 March 2023 10:40:47
18 March 2023 10:40:47

#615 [——€] PS7 ~\Documents\PowerShell>
```   

 (2) if you run something like `Get-Date` | format-table | out-file delete.me  How many leading and trailing carriage returns do we expect in the file, and what should be done to ensure they are added or not ? To be consistent a lot of other commands would need specifiers for no leading break  (`$d, $d.AddDays(1) > delete.me` produces  file which starts with a blank line) as well as no trailing break or only single final break etc.  Yes, in isolation `$String1 | out-string | out-string | out-string` adding a line break every time looks daft. But it is how `$string1 > delete.me ` followed by `$string2 >> delete.me` avoids mashing lines together. 

(3) We have had the present formatting with accepted inconsistencies for a long time. We know that whatever it does by default will not work in every case. Work rounds are understood and some people will have functions which take a string from Out-string and reformat it for their use case. Adding a switch to support an exception works for the person bothered by that exception. Speaking for myself, I'd like `out-string` to be culture aware so I could do `get-date | out-string -culture fr-fr` because using local culture is sometimes wrong. If every exceptional case was given a switch it would be never-ending.  And of course, old scripts would work with the new version but as soon as someone takes a script which uses the new switch and tries to run it on a server where only Windows PowerShell exists things break, and there is a quite a high bar set for doing that.  

My hunch is questions in this area will continue to be asked, and the door is not nailed shut on this, but hopefully that explains why for this case, now, the answer is no.

**Comment by mklement0** [comment]
> A new switch to strip the trailing new line wouldn't break anything, and is a suggested resolution in the OP.

Good point, @ImportTaste. I hope that

* #5108

(which covers multiple cmdlets, not just `0ut-String`) will now be considered for implementation.

**Comment by mklement0** [comment]
> The issue is more complicated than posed in the OP.

The quirks and complexities of the formatting system are a red herring with respect to what this issue asked for.

It is a simple as it gets:

Switch from:

```powershell
(<whatever-lines-the-formatting-system-generates> -join [Environment]::NewLine) + [Environment]::NewLine
```

to just:

```powershell
<whatever-lines-the-formatting-system-generates> -join [Environment]::NewLine
```

**That extra newline was never justified - but whether or not it is present _doesn't matter if you use `Out-String` for actual (complex) object formatting_** (say, `Get-ChildItem` output), because there you decidedly get a _for-display-only_ representation whose specifics you shouldn't rely on for _programmatic processing_.

If you conceive of `Out-String` as cmdlet _solely_ for emitting _for-display_ representations - and it certainly _primarily_ is -  then changing its output would _not_ be a backward-compatibility concern, because _for-display_ representations aren't part of the backward-compatibility contract.

**In (widespread) practice, however, `Out-String` is _also_ used to emit _data_**, namely in calls to _native (external) programs_, **as a convenient way of getting an external program's output  as a _single, multiline string_** (instead of the line-by-line streaming that happens by default).

Given that external-program output is only ever interpreted as _strings_, and strings render as-is (in formatting-system speak, they render _out of band_), `Out-String` output - which then amounts to:

```powershell
(<lines-from-external-program> -join [Environment]::NewLine) + [Environment]::NewLine
 ```

_is_ suitable for programmatic processing  - but **in _that_ case the trailing newline is decidedly a nuisance**:

* If you split the string into lines later (`-split '\r?\n'`), you'll get an _extra, empty array element_.

* If you send it to `Out-File` or `Set-Content` later, you'll get an extra empty line.

**Comment by CyberBorga** [comment]
I love Powershell, but as a long time Linux bash user, **I agree with @mklement0**.  I'd prefer if Out-String cmdlet did NOT add a trailing newline character (i.e., if I wanted a newline, I would have told you to add one, or I would have added one myself!).  

However, since Powershell commands work with objects (vs. bash commands which usually work with text), a **possible answer** lies in Powershell's **object-oriented inheritance**.  Since all objects inherit their methods and properties from their parents, I would recommend using the ToString method available to all .NET/Powershell objects.  

In the example above, you can use ToString to convert the pipeline to strings vs. using the Out-String cmdlet:

`1, 2 | % { $_.ToString()}`

This should produce the results you wanted:

```
PS C:\Users\cyberborga> 1, 2 | % { $_.ToString()}
1
2
PS C:\Users\cyberborga>
```

Of course, since child objects can override the ToString method, your results may be unpredictable for more complex objects, such as the System.DateTime object returned by Get-Date, which prints the date/time in "the current culture's short date and long time pattern".

```
PS C:\Users\cyberborga> Get-Date  | % { $_.ToString()}
3/17/2023 9:42:31 AM
PS C:\Users\cyberborga>
```

In these cases, I think the easiest answer is just to ask the cmdlet to output the results in a format you would like (vs. using the IFormatProvider options).  

In the example above, you could use:

PS C:\Users\cyberborga> Get-Date -Format "dddd, MMMM, d, yyyy hh:mm:ss tt"
Friday, March, 17, 2023 09:54:07 AM
PS C:\Users\cyberborga>

* NOTE: You don't need to convert the results of this command to a string, because the Get-Date cmdlet outputted a string.

I hope this helps.

**Comment by CyberBorga** [comment]
Of course, an even easy answer might be to just use the Out-String -Stream option, which would preserve the "normal" output of cmdlets (i.e., Get-Date wouldn't require any extra formatting to get the normal 1> outputs):

```
 PS C:\Users\cyberborga> Get-Date

Friday, March 17, 2023 10:12:46 AM


PS C:\Users\cyberborga> Get-Date | Out-String -Stream

Friday, March 17, 2023 10:12:55 AM


PS C:\Users\cyberborga>
```

Good luck.

**Comment by jhoneill** [comment]
@CyberBorga  however -stream outputs as many strings as there are objects, so to they may need to be joined together.

**Comment by mklement0** [comment]
@CyberBorga, to flesh out @jhoneill's comment:

* External-program output _implicitly_ behaves like `Out-String -Stream`: that is, stdout is broken into _lines_ (of what is invariably treated as _string_ output) and each line is streamed to the pipeline as it becomes available. Therefore,
`& $someExternalCli … | % ToString` and 
`& $someExternalCli … | Out-String -Stream` (note that there's also a built-in `oss` function that wraps `Out-String -Stream`) are in effect expensive no-ops, because they behave the same as just
`& $someExternalCli …`

* The use case that this proposal targets is `Out-String`'s _default_ behavior of emitting a _single, multiline string_ rather than a a stream of lines (which in turn into an _array_ of strings when captured).

* As such, the workaround is `(& $someExternalCli …) -join [Environment]::NewLine` - not terrible, but less obvious, more complex, and less PowerShell-idiomatic than `& $someExternalCli … | Out-String`

**Comment by TheOnlyWei** [comment]
> The Cmdlewt working group has reviewed this issue. We have decided changing the default behavior would break way too many things. Work rounds for things that don't like trailing line breaks have be around for a long time. Resolved as by design.

If it is "by design", then the design should be reconsidered. Resolution by design should be for cases when the design is sound, which in this case, I don't think it is good design to modify user string. A "string" does not mean "a bunch of characters that end with a newline character". Maybe rename the function to "Out-StringWithNewline".

**Comment by jhoneill** [comment]
@TheOnlyWei  

The job `Out-String` was designed to do is to **output the formatted text for...** which includes.  
1. producing formatted strings from the format objects output by `format-*` commands to writable text printing or writing into a file.   So you can do `$Table = get-process P* | format-table -autosize` ;     
and then `$table | Out-String -width 240 > textfile` ; or `$table | out-String -width 60` 
2. Detecting objects which should be formatted with `format-*`  and producing the objects for (1) ...   

Single new-lines, and double new lines are printed based on what seemed like a good idea at the time to the original designers, and that has  become 'established custom and practice' meaning changing them would please some people and break code which sees 'custom and practice' as an absolute guarantee of behaviour. 

`$x | Out-String  `  is **emphatically not** the same as `[string]$x`  or `$x.tostring()`

for example 
``` 
PS> $x = 'a', 'b', 'c'
PS> $x.ToString()
System.Object[]
PS> [string]$x
a b c
PS> $x | Out-String
a
b
c

PS> $x -join [System.Environment]::NewLine
a
b
c
PS>
```

**None of those is right on all occasions.** 

> A "string" does not mean "a bunch of characters that must end with a newline character".

Indeed.  String is any bunch of characters. We do have zero terminated strings, but .NET strings aren't. The "formatted output for" probably  ends with a new line, but might end with more than one... 

> It wouldn't be an issue if the function was named "Out-StringWithNewline". 

'String' in the name should be read as  "String formatted for printing as is".  Picking  Out-FormattedString as the name all those years ago might have saved confusion :-)   But the _String_ for 1234 and the _string formatted for printing_ 1234 are different meaning that     
`1234 | out-string | format-hex`  
and 
` (1234).tostring() | format-hex`        
Show different things.  

> This caveat will continue cost dev time for users because of the surprise newline appended to the end.

**Yes.** As long as people do not understand that `Out-String` is making a formatted print-ready version and is distinct from "the string representation of" _time will be lost._ 
I've shown three ways to make an array into a single string, all valid, and **all three are  _potentially_ surprising.** 
How confident is a reasonably experienced user certain what   `"my array is   $x "   ` and `"my array is  {0} " -f $x `  will do?    
If $d is a date does that user know when `"The date is " + $d`  is different from `"The date is " + $d.ToString()` ? 
(The former does a culture invariant cast, but `ToString` is culture aware).  As a non-American I **HATE** that.  

**Any ways of helping people understand and get to the version they want more quickly will be gratefully received.** 
But could we avoid arguments about 15 year old decisions on formatting, please?. The empirical evidence is they has been sufficiently OK, for enough of people, enough of the time, and a different decision would probably have roughly as good/bad,  annoying a different set of people on different occasions. I accept that I'll never see a change to make dates culture aware where **I** think they should be  'it has always been that way' is the worst reason, and yet the change I'd like would break things which rely on established practice. With 15 years of PowerShell behind me I **still** get surprised by date behaviour so I have some sympathy for anyone who says "There are three ways to convert to string and none is what I want/expect this operation to do!". 

There **is** a discussion point about whether "Produce the formatted print-ready version BUT WITHOUT THE TRAILING LINE BREAK"  is a such a desirable scenario that it should be incorporated into `Out-String`.   The review that happened said no, as I said above that could be revisited.  People do believe passionately that decisions made long ago weren't OK, but sometimes that yields arguments which should be left for student moots, or debating societies.  Arguing that naming the cmdlet `out-string` leads to wrong expectations of what it does, or that some formatting it does isn't optimal both fall into that category - at least in my view. On the other hand saying "I'd really benefit from the a `-NoTrailingNewLine` switch because... " can help us to see things that should be reconsidered (and also applies when if the old decision was perfect).

**Comment by JEmlay** [comment]
Wow, 2 years with no resolution.  Good job Powershell team.
